crowbar_framework/lib/crowbar/chef/upload.rb

```diff
@@ -0,0 +1,284 @@
+# You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require "chef/cookbook_uploader"
+require "chef/knife/core/object_loader"
+
+module Crowbar
+  class Chef
+    class Upload
+      class << self
+        def all
+          cookbooks
+          data_bags
+          roles
+        end
+
+        def cookbooks
+          version_constraints_to_update = Hash.new
+          cookbooks_for_upload = Array.new
+
+          cookbook_loader.each do |cookbook_name, cookbook|
+            cookbooks_for_upload.push(cookbook)
+            version_constraints_to_update[cookbook_name] = cookbook.version
+          end
+
+          logger.info("Checking existing cookbooks...")
+          cookbooks_for_upload.reject! do |cookbook|
+            next unless api_coobbooks.include?(cookbook.name.to_s)
+            cookbook_exists?(cookbook)
+          end
+
+          unless cookbooks_for_upload.any?
+            logger.info("All cookbooks are already existing on the server")
+            return true
+          end
+
+          cookbooks_for_upload.each do |cookbook|
+            logger.info("Checking cookbook #{cookbook.name}...")
+            return false unless check_links(cookbook)
+            return false unless check_dependencies(cookbook)
+          end
+
+          logger.info("Uploading cookbooks... #{cookbooks_for_upload.map(&:name).join(", ")}")
+          uploader = ::Chef::CookbookUploader.new(
+            cookbooks_for_upload,
+            force: false,
+            concurrency: cookbooks_for_upload.count
+          )
+          return false unless uploader.upload_cookbooks
+
+          true
+        end
+
+        def data_bags
+          loader = ::Chef::Knife::Core::ObjectLoader.new(::Chef::DataBagItem, logger)
+          data_bags = loader.find_all_object_dirs(chef_data_bags_path) || []
+          data_bags.each do |data_bag|
+            data_bag_items = loader.find_all_objects(chef_data_bags_path.join(data_bag))
+            data_bag_item_paths = normalize_data_bag_item_paths(data_bag_items) || []
+            data_bag_item_paths.each do |data_bag_item_path|
+              # Workaround for a strange chef behavior
+              relative_path = chef_data_bags_path.relative_path_from(Pathname.new(Dir.pwd))
+              data_bag_item = loader.load_from(relative_path, data_bag, data_bag_item_path)
+              bag = ::Chef::DataBagItem.new
+              bag.data_bag(data_bag)
+              bag.raw_data = data_bag_item
+              logger.info("Uploading data_bag #{data_bag_item_path}")
+
+              return false unless bag.save
```

Comment:
I think `bag.save` either returns a `DataBagItem` or raises a `Net::HTTPServerException` (at least in the newer chef) so maybe we have to treat this differently here.
http://www.rubydoc.info/github/opscode/chef/Chef/DataBagItem#save-instance_method

can you please double check?